src/bases.jl

```diff
@@ -68,96 +70,114 @@ Available methods:
+    return DiagonalizingOrthonormalBasis{typeof(X),𝔽}(X)
+end
+struct DiagonalizingBasisData{D,V,ET}
+    frame_direction::D
+    eigenvalues::ET
+    vectors::V
 end
-
-const ArbitraryOrDiagonalizingBasis =
-    Union{ArbitraryOrthonormalBasis,DiagonalizingOrthonormalBasis}
 
 """
-    PrecomputedOrthonormalBasis(vectors::AbstractVector, field::AbstractNumbers = ℝ)
+    ProductBasisData
 
-A precomputed orthonormal basis at a point on a manifold.
-
-The type parameter `field` denotes the [`AbstractNumbers`](@ref) that will be used as
-scalars.
+A typed tuple to store tuples of data of stored/precomputed bases.
 """
-struct PrecomputedOrthonormalBasis{TV<:AbstractVector,𝔽} <:
-       AbstractPrecomputedOrthonormalBasis{𝔽}
-    vectors::TV
+struct ProductBasisData{T<:Tuple}
+    parts::T
 end
 
-function PrecomputedOrthonormalBasis(vectors::AbstractVector, field::AbstractNumbers = ℝ)
-    return PrecomputedOrthonormalBasis{typeof(vectors),field}(vectors)
-end
 
-@doc raw"""
-    DiagonalizingOrthonormalBasis(vectors, kappas, field::AbstractNumbers = ℝ)
+const ArbitraryOrDiagonalizingBasis =
+    Union{ArbitraryOrthonormalBasis,DiagonalizingOrthonormalBasis}
 
-A precomputed orthonormal basis `Ξ` as a vector of tangent vectors (of length determined
-by [`manifold_dimension`](@ref)) in the tangent space that diagonalizes the curvature
-tensor $R(u,v)w$ with eigenvalues `kappas` and where the direction `v` has eigenvalue `0`.
 
-The type parameter `field` denotes the [`AbstractNumbers`](@ref) that will be used as
-scalars.
-"""
-struct PrecomputedDiagonalizingOrthonormalBasis{TV<:AbstractVector,TK<:AbstractVector,𝔽} <:
-       AbstractPrecomputedOrthonormalBasis{𝔽}
-    vectors::TV
-    kappas::TK
-end
-
-function PrecomputedDiagonalizingOrthonormalBasis(
-    vectors::AbstractVector,
-    kappas::AbstractVector,
-    field::AbstractNumbers = ℝ,
-)
-    return PrecomputedDiagonalizingOrthonormalBasis{typeof(vectors),typeof(kappas),field}(
-        vectors,
-        kappas,
-    )
+struct CachedBasis{B,V,𝔽} <: AbstractBasis{𝔽} where {BT<:AbstractBasis, V}
+    data::V
+end
+function CachedBasis(basis::B, data::V, 𝔽::AbstractNumbers = ℝ) where  {V,B<:AbstractBasis}
+    return CachedBasis{B,V,𝔽}(data)
+end
+function CachedBasis(basis::CachedBasis) # avoid double encapsulation
+    return basis
+end
+function CachedBasis(
+    basis::DiagonalizingOrthonormalBasis,
+    eigenvalues::ET,
+    vectors::T,
+    𝔽::AbstractNumbers = ℝ,
+) where {ET<:AbstractVector, T<:AbstractVector}
+    data = DiagonalizingBasisData(basis.frame_direction, eigenvalues, vectors)
+    return CachedBasis(basis, data, 𝔽)
 end
 
+function _euclidean_basis_vector(p, i)
+    X = zero(p)
+    X[i] = 1
+    return X
+end
+
+_get_vectors(B::CachedBasis) = B.data
+_get_vectors(B::CachedBasis{BT,D,𝔽}) where {BT<:AbstractBasis, D<:DiagonalizingBasisData, 𝔽} = B.data.vectors
+_get_vectors(B::CachedBasis{BT,D,𝔽}) where {BT<:AbstractBasis, D<:ProductBasisData, 𝔽} = B.data.parts
```

Comment:
As far as I can see this function is called by `get_vectors` which is supposed to return a `Vector` of tangent vectors. `data.parts` is a very different thing.

Reply:
The naming is not great, it's more like an internal extract.

Reply:
Well, here: https://github.com/JuliaNLSolvers/Manifolds.jl/pull/115/files#diff-9dbb0d28119f714ea432396713884380R348 it's used to return vectors outside.

Reply:
Thanks for noticing, then I have to check how that came into the code and what it actually should be.

Reply:
The idea for the internal function is, that that can be used in show. But now I split them more carefully and it only returns vectors when its really array of vectors. I now just have to get the tests back up running. Problem stays: Someone has to redo the shows for cached data. I am not yet able to understand those tricky functions.

Reply:
I'll take a look at the show methods for cached data in a moment.